moveit_ros/visualization/handeye_calibration_rviz_plugin/src/handeye_control_widget.cpp

```diff
@@ -0,0 +1,828 @@
+    }
+    catch (pluginlib::PluginlibException& ex)
+    {
+      QMessageBox::warning(this, tr("Exception while creating handeye solver plugin loader "), tr(ex.what()));
+      return false;
+    }
+  }
+
+  // Get available plugins
+  plugins = solver_plugins_loader_->getDeclaredClasses();
+  return !plugins.empty();
+}
+
+bool ControlTabWidget::createSolverInstance(const std::string& plugin_name)
+{
+  try
+  {
+    solver_ = solver_plugins_loader_->createUniqueInstance(plugin_name);
+    solver_->initialize();
+  }
+  catch (pluginlib::PluginlibException& ex)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Exception while loading handeye solver plugin: " << plugin_name << ex.what());
+    solver_ = nullptr;
+  }
+
+  return solver_ != nullptr;
+}
+
+void ControlTabWidget::fillSolverTypes(const std::vector<std::string>& plugins)
+{
+  for (const std::string& plugin : plugins)
+    if (!plugin.empty() && createSolverInstance(plugin))
+    {
+      const std::vector<std::string>& solvers = solver_->getSolverNames();
+      for (const std::string& solver : solvers)
+      {
+        std::string solver_name = plugin + "/" + solver;  // solver name format is "plugin_name/solver_name"
+        calibration_solver_->addItem(tr(solver_name.c_str()));
+      }
+    }
+}
+
+std::string ControlTabWidget::parseSolverName(const std::string& solver_name, char delimiter)
+{
+  std::vector<std::string> tokens;
+  std::string token;
+  std::istringstream tokenStream(solver_name);
+  while (std::getline(tokenStream, token, delimiter))
+  {
+    tokens.push_back(token);
+  }
+  return tokens.back();
+}
+
+bool ControlTabWidget::takeTranformSamples()
+{
+  // Store the pair of two tf transforms and calculate camera_robot pose
+  try
+  {
+    geometry_msgs::TransformStamped cTo;
+    geometry_msgs::TransformStamped bTe;
+
+    // Get the transform of the object w.r.t the camera
+    cTo = tf_buffer_->lookupTransform(frame_names_["sensor"], frame_names_["object"], ros::Time(0));
+
+    // Get the transform of the end-effector w.r.t the robot base
+    bTe = tf_buffer_->lookupTransform(frame_names_["base"], frame_names_["eef"], ros::Time(0));
+
+    // save the pose samples
+    effector_wrt_world_.push_back(tf2::transformToEigen(bTe));
+    object_wrt_sensor_.push_back(tf2::transformToEigen(cTo));
+
+    ControlTabWidget::addPoseSampleToTreeView(cTo, bTe, effector_wrt_world_.size());
+  }
+  catch (tf2::TransformException& e)
+  {
+    ROS_WARN("TF exception: %s", e.what());
+    return false;
+  }
+
+  return true;
+}
+
+bool ControlTabWidget::solveCameraRobotPose()
+{
+  if (solver_ && !calibration_solver_->currentText().isEmpty())
+  {
+    bool res = solver_->solve(effector_wrt_world_, object_wrt_sensor_, sensor_mount_type_,
+                              parseSolverName(calibration_solver_->currentText().toStdString(), '/'));
+    if (res)
+    {
+      camera_robot_pose_ = solver_->getCameraRobotPose();
```

Comment:
If this returns all 0, or the solver returns a failure, an error should be displayed here.

Reply:
In my branch I've added a warning if the solver fails (PR to come). By "returns all 0", do you mean that the pose is the identity? If the result is the identity, that definitely suggests that something went wrong, but at the same time it is a valid result. I'd rather work with the expectation that the return values of `solve` can be trusted, and so there's no need to double-check the solution here.

Reply:
I meant to check for solver failure, that's all. I wrote it this way because if the pose is not filled in this line, that would imply that the solver failed.